File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.cpp
class Solution {
public:
    int search(vector<int>& nums, int k) {
        int n = nums.size();
      int l=0, r=n-1;

      while(l<=r){
        int mid = l + (r-l)/2;

        if(nums[mid] == k)
            return mid;

        else if(nums[l] == nums[r] && nums[r] == nums[mid] && nums[l] == nums[mid]){
            l++;
            r--;
        }

        else if(nums[l] <= nums[mid]){
            if(nums[l] <= k && k < nums[mid])
                r = mid-1;
            else
                l = mid+1;
        }

        else if(nums[mid] <= nums[r]){
            if(nums[mid] < k && k <= nums[r])
                l = mid+1;
            else
                r = mid-1;
        }
      }
      return -1;
    }
};

File: Data Structures & Algorithms/permutations/submission-0.java
class Solution {

    void solve(boolean[] vis, int[] nums, List<Integer> tmp, List<List<Integer>> ans){
        if(tmp.size() == nums.length){
            ans.add(new ArrayList<>(tmp));
            return;
        }

        for(int i=0; i<nums.length; i++){
            if(vis[i])
                continue;
            vis[i] = true;
            tmp.add(nums[i]);
            solve(vis, nums, tmp, ans);
            tmp.remove(tmp.size()-1);
            vis[i] = false;
        }
    }

    public List<List<Integer>> permute(int[] nums) {
        List<Integer> tmp = new ArrayList<>();
        List<List<Integer>> ans = new ArrayList<>();
        
        boolean[] vis = new boolean[nums.length];

        solve(vis, nums, tmp, ans);
        return ans;
    }
}
